File: curriculum.py
class Node:
    __slots__ = "son", "end"
    def __init__(self):
        self.son = [None] * 26
        self.end = False

class Trie:
    def __init__(self):
        self.root = Node()
    
    
    def insert(self, word: str) -> None:
        cur = self.root
        for c in word:
            c = ord(c) - ord('a')
            if cur.son[c] is None:
                cur.son[c] = Node()
            cur = cur.son[c]
        cur.end = True
        
    
    def find(self, word: str) -> int:
        cur = self.root
        for c in word:
            c = ord(c) - ord('a')
            if cur.son[c] is None:
                return 0
            cur = cur.son[c]
        return 2 if cur.end else 1
    
    def search(self, word: str) -> bool:
        return self.find(word) == 2
            
            
        
        
    def startsWith(self, prefix: str) -> bool:
        return self.find(prefix) != 0
    

File: curriculum.java
class Trie {
    class Node {
        Node[] son = new Node[26];
        boolean end = false;
    }

    private final Node root = new Node();


    public void insert(String word){
        Node cur = root;
        for (char c: word.toCharArray()){
            c -= 'a';
            if(cur.son[c] == null){
                cur.son[c] = new Node();
            }
            cur = cur.son[c];
        }
        cur.end = true;

    }

    public boolean search(String word){
        return find(word) == 2;
    }

    public boolean startsWith(String prefix){
        return find(prefix) != 0;
    }


    public int find(String word){
        Node cur = root;
        for(char c: word.toCharArray()){
            c -= 'a';
            if(cur.son[c] == null){
                return 0;
            }
            cur = cur.son[c];
        }

        return cur.end ? 2 : 1;
    }
}